modules/dev-utils/src/main/java/org/apache/ignite/development/utils/IgniteIndexReader.java

```diff
@@ -0,0 +1,2130 @@
+            int lvlsCnt = metaIO.getLevelsCount(addr);
+
+            long[] firstPageIds = IntStream.range(0, lvlsCnt).mapToLong(i -> metaIO.getFirstPageId(addr, i)).toArray();
+
+            for (int i = 0; i < lvlsCnt; i++) {
+                long pageId = firstPageIds[i];
+
+                while (pageId > 0) {
+                    try {
+                        buf.rewind();
+
+                        readPage(store, pageId, buf);
+
+                        pageIO = PageIO.getPageIO(addr);
+
+                        if (i == 0 && !(pageIO instanceof BPlusLeafIO))
+                            throw new IgniteException("Not-leaf page found on leaf level, pageId=" + pageId + ", level=" + i);
+
+                        if (!(pageIO instanceof BPlusIO))
+                            throw new IgniteException("Not-BPlus page found, pageId=" + pageId + ", level=" + i);
+
+                        ioStat.computeIfAbsent(pageIO.getClass(), k -> new AtomicLong(0)).incrementAndGet();
+
+                        if (pageIO instanceof BPlusLeafIO) {
+                            PageIOProcessor ioProcessor = getIOProcessor(pageIO);
+
+                            PageContent pageContent = ioProcessor.getContent(pageIO, addr, pageId, treeCtx);
+
+                            pageContent.items.forEach(itemStorage::add);
+                        }
+
+                        pageId = ((BPlusIO)pageIO).getForward(addr);
+                    }
+                    catch (Throwable e) {
+                        errors.computeIfAbsent(pageId, k -> new LinkedList<>()).add(e);
+
+                        pageId = 0;
+                    }
+                }
+            }
+        }
+        catch (Throwable e) {
+            errors.computeIfAbsent(rootPageId, k -> new LinkedList<>()).add(e);
+        }
+        finally {
+            freeBuffer(buf);
+        }
+
+        return new TreeTraversalInfo(ioStat, errors, null, rootPageId, itemStorage);
+    }
+
+    /**
+     * Gets tree node and all its children.
+     *
+     * @param pageId Page id, where tree node is located.
+     * @param nodeCtx Tree traverse context.
+     * @return Tree node.
+     */
+    private TreeNode getTreeNode(long pageId, TreeTraverseContext nodeCtx) {
+        PageContent pageContent;
+
+        PageIOProcessor ioProcessor;
+
+        try {
+            final ByteBuffer buf = allocateBuffer(pageSize);
+
+            try {
+                readPage(nodeCtx.store, pageId, buf);
+
+                final long addr = bufferAddress(buf);
+
+                final PageIO io = PageIO.getPageIO(addr);
+
+                nodeCtx.ioStat.computeIfAbsent(io.getClass(), k -> new AtomicLong(0)).incrementAndGet();
+
+                ioProcessor = getIOProcessor(io);
+
+                pageContent = ioProcessor.getContent(io, addr, pageId, nodeCtx);
+            }
+            finally {
+                freeBuffer(buf);
+            }
+
+            return ioProcessor.getNode(pageContent, pageId, nodeCtx);
```

Comment:
It looks too strange:
pageContent = ioProcessor.getContent(io, addr, pageId, nodeCtx);
ioProcessor.getNode(pageContent, pageId, nodeCtx);
You are getting some object from other and returning it to itself.

Reply:
Maybe it does, but it is designed in such way to free the buffer before recursive call, i.e. it's splitting one operation to two steps, the buffer can be deallocated after first one.

Reply:
Why does not free buffer after last one?

Reply:
In some cases we need only PageContent without creating TreeNode (see #horizontalTreeScan)